lib/puppet/application/apply.rb

```diff
@@ -350,18 +310,98 @@ def setup
   private
 
   def read_catalog(text)
-    format = Puppet::Resource::Catalog.default_format
-    begin
-      catalog = Puppet::Resource::Catalog.convert_from(format, text)
-    rescue => detail
-      raise Puppet::Error, _("Could not deserialize catalog from %{format}: %{detail}") % { format: format, detail: detail }, detail.backtrace
-    end
+    facts = get_facts()
+    node = get_node()
+    configured_environment = get_configured_environment(node)
+
+    # TRANSLATORS "puppet apply" is a program command and should not be translated
+    Puppet.override({:current_environment => configured_environment}, _("For puppet apply")) do
+      configure_node_facts(node, facts)
+
+      # NOTE: Does not set rich_data = true automatically (which would ensure always reading catalog with rich data
+      # on (seemingly the right thing to do)), but that would remove the ability to test what happens when a
+      # rich catalog is processed without rich_data being turned on.
+      format = Puppet::Resource::Catalog.default_format
+      begin
+        catalog = Puppet::Resource::Catalog.convert_from(format, text)
+      rescue => detail
+        raise Puppet::Error, _("Could not deserialize catalog from %{format}: %{detail}") % { format: format, detail: detail }, detail.backtrace
+      end
+      # Resolve all deferred values and replace them / mutate the catalog
+      Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(node, node.facts, catalog)
```

Comment:
The signature for this method was changes to allow for cached_catalogs to work in #7022, so it should just be `(node.facts, catalog)`.

Reply:
fixed